### PR TITLE
REF: handle where blocks in tuple <-> struct conversion

### PR DESCRIPTION
**Pull request by Kobzol**
Fixes: https://github.com/intellij-rust/intellij-rust/issues/6693

changelog: Correctly move where blocks when converting between structs and tuples.

### CODE REVIEW

File: src/test/kotlin/org/rust/ide/refactoring/ConvertToNamedFieldsRefactoringTest.kt [hunk 1]
```diff
@@ -182,6 +182,14 @@ class ConvertToNamedFieldsRefactoringTest : RsTestBase() {
         }
     """)
 
+    fun `test where clause`() = doAvailableTest("""
+        trait Trait {}
+        struct Test/*caret*/<T>(T) where T: Trait ;
```
Comment (Kobzol) [inline src/test/kotlin/org/rust/ide/refactoring/ConvertToNamedFieldsRefactoringTest.kt:187]: The space after `Trait` is deliberate, to test that the whitespace is removed.

File: src/test/kotlin/org/rust/ide/refactoring/ConvertToTupleRefactoringTest.kt [hunk 2]
```diff
@@ -142,6 +142,16 @@ class ConvertToTupleRefactoringTest : RsTestBase() {
...
+    fun `test where clause`() = doAvailableTest("""
+        trait Trait {}
+        struct Test<T> where T: Trait {/*caret*/
+            a: T,
+        }
+    """, """
+        trait Trait {}
+        struct Test<T>(T) where T: Trait;
+    """)
+
```
Comment (dima74) [inline src/test/kotlin/org/rust/ide/refactoring/ConvertToTupleRefactoringTest.kt:154]:
```suggestion

    fun `test where clause (multiline)`() = doAvailableTest("""
        trait Trait {}
        struct Test<T1, T2>
            where
                T1: Trait,
                T2: Trait,
        {/*caret*/
            a: T1,
            b: T2,
        }
    """, """
        trait Trait {}
        struct Test<T1, T2>(T1, T2)
            where
                T1: Trait,
                T2: Trait;
    """)
```

File: src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt [hunk 3]
```diff
@@ -110,8 +112,16 @@ class RsConvertToTupleProcessor(project: Project, val element: RsFieldsOwner, va
...
+        element.blockFields?.replace(newTuplePsiElement.tupleFields!!)
+        if (element is RsStructItem) {
+            var anchor: PsiElement? = element.tupleFields
+            val whereClause = element.whereClause
+            if (whereClause != null) {
+                anchor = element.addAfter(whereClause, element.tupleFields)
+                whereClause.deleteWithSurroundingWhitespace()
+            }
+            element.addAfter(rsPsiFactory.createSemicolon(), anchor)
+        }
```
Comment (dima74) [inline src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt:124]: If `where` clause is multiline, we need to keep `where` on new line. For that reason probably it is worth to keep where clause as-is and instead move block/tuple fields, something like this:

```kotlin
val blockFields = element.blockFields ?: return
val tupleFields = newTuplePsiElement.tupleFields ?: return
val whereClause = (element as? RsStructItem)?.whereClause
if (whereClause == null) {
    element.blockFields!!.replace(tupleFields)
} else {
    element.addAfter(tupleFields, whereClause.getPrevNonWhitespaceSibling())
    (blockFields.prevSibling as? PsiWhiteSpace)?.delete()
    blockFields.delete()
    whereClause.lastChild.takeIf { it.elementType == COMMA }?.delete()
}
if (element is RsStructItem) element.addAfter(rsPsiFactory.createSemicolon(), element.lastChild)
```

Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/refactoring/convertStruct/RsConvertToTupleProcessor.kt:124]: Awesome. Thanks!

**Comment by dima74** [comment]
Please also handle multiline `where` clause in `RsConvertToNamedFieldsProcessor`

**Comment by dima74** [comment]
Could you please squash commits

**Comment by dima74** [comment]
Thanks!
bors r+

**Pull request merged**